libnetdata/popen/popen.c

```diff
@@ -78,36 +78,75 @@ static void myp_del(pid_t pid) {
 #define PIPE_READ 0
 #define PIPE_WRITE 1
 
-/* custom_popene flag definitions */
-#define FLAG_CREATE_PIPE    1 // Create a pipe like popen() when set, otherwise set stdout to /dev/null
-#define FLAG_CLOSE_FD       2 // Close all file descriptors other than STDIN_FILENO, STDOUT_FILENO, STDERR_FILENO
-
 /*
- * Returns -1 on failure, 0 on success. When FLAG_CREATE_PIPE is set, on success set the FILE *fp pointer.
+ * Returns -1 on failure, 0 on success. When POPEN_FLAG_CREATE_PIPE is set, on success set the FILE *fp pointer.
  */
-static inline int custom_popene(const char *command, volatile pid_t *pidptr, char **env, uint8_t flags, FILE **fpp) {
+int custom_popene_internal_dont_use_directly(volatile pid_t *pidptr, char **env, uint8_t flags, FILE **fpp, const char *command, ...) {
+    // convert the variable list arguments into what posix_spawn() needs
+    // all arguments are expected strings
```

Comment:
Is there any added advantage of using variadic arguments in this scenario instead of simple array of strings?

Reply:
well, we have to support variadic parameters somehow, so that at other places we can write:

```c
popen(command, param1, param2, param3, param4, ...);
```

If it is not in this function, it will be on another that will convert variadic parameters to an array of strings to call this function.

I don't think it is practical to ask for the callers to always prepare an array of strings. A lot of work without reason...

Reply:
This is why I created the `mypopen_raw()` macros. To hide this complexity.